Add support for empty replica pages

denesb opened the pull request:
Many tombstones in a partition is a problem that has been plaguing queries since the inception of Scylla (and even before that as they are a pain in Apache Cassandra too). Tombstones don't count towards the query's page limit, neither the size nor the row number one. Hence, large spans of tombstones (be that row- or range-tombstones) are problematic: the query can time out while processing this span of tombstones, as it waits for more live rows to fill the page. In the extreme case a partition becomes entirely unreadable, all read attempts timing out, until compaction manages to purge the tombstones.
The solution proposed in this PR is to pass down a tombstone limit to replicas: when this limit is reached, the replica cuts the page and marks it as short one, even if the page is empty currently. To make this work, we use the last-position infrastructure added recently by 3131cbea62, so that replicas can provide the position of the last processed item to continue the next page from. Without this no forward progress could be made in the case of an empty page: the query would continue from the same position on the next page, having to process the same span of tombstones.
The limit can be configured with the newly added `query_tombstone_limit` configuration item, defaulted to 10000. The coordinator will pass this to the newly added `tombstone_limit` field of `read_command`, if the `replica_empty_pages` cluster feature is set.

Upgrade sanity test was conducted as following:
* Created cluster of 3 nodes with RF=3 with master version
* Wrote small dataset of 1000 rows.
* Deleted prefix of 980 rows.
* Started read workload: `scylla-bench -mode=read -workload=uniform -replication-factor=3 -nodes 127.0.0.1,127.0.0.2,127.0.0.3 -clustering-row-count=10000 -duration=10m -rows-per-request=9000 -page-size=100`
* Also did some manual queries via `cqlsh` with smaller page size and tracing on.
* Stopped and upgraded each node one-by-one. New nodes were started by `--query-tombstone-page-limit=10`.
* Confirmed there are no errors or read-repairs.

Perf regression test:
```
build/release/test/perf/perf_simple_query_g -c1 -m2G --concurrency=1000 --task-quota-ms 10 --duration=60
```
Before:
```
median 133665.96 tps ( 62.0 allocs/op,  12.0 tasks/op,   43007 insns/op,        0 errors)
median absolute deviation: 973.40
maximum: 135511.63
minimum: 104978.74
```
After:
```
median 129984.90 tps ( 62.0 allocs/op,  12.0 tasks/op,   43181 insns/op,        0 errors)
median absolute deviation: 2979.13
maximum: 134538.13
minimum: 114688.07
```
Diff: +~200 instruction/op.

Fixes: https://github.com/scylladb/scylla/issues/7689
Fixes: https://github.com/scylladb/scylla/issues/3914
Fixes: https://github.com/scylladb/scylla/issues/7933
Refs: https://github.com/scylladb/scylla/issues/3672

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1398/

avikivity commented:
> TODO:
> 
> * Handle mismatching continue positions on the coordinator: if the digest matches, just pick the smallest one as the continue position.

Sounds simple.

> * Multi-node tests, with various mismatching amount of tombstones on different nodes.
> * [optimization] allow querier lookup to succeed if continue position is < then querier's position, if we know that the range `[$continue_pos, $querier_pos]` contains only dead data.


We should only do this if it's a problem in practice (and we have counters to see if it's a problem or not).

> 
> Questions:
> 
> * We already have a `tombstone_{warn,fail}_threshold` config items, inherited from C*, currently unused. Should we use these instead of adding a new item. My main problem with this is the name: `tombstone_fail_threshold` doesn't really tell a user that this just means the page is cut (it is accurate for unpaged queries though).

Certainly we should add a new item (if a new configuration is really warranted).

avikivity commented on idl/query.idl.hh:
```diff
@@ -27,7 +27,6 @@ class qr_partition stub [[writable]] {
 class query_result stub [[writable]] {
     utils::chunked_vector<qr_partition> partitions; // in ring order
-    std::optional<full_position> last_position [[version 5.1]];
 };
```
Please extract this commit into its own pull so we can commit it quickly and not accidentally cut a branch while this is making progress.

avikivity commented on query-request.hh:
```diff
@@ -261,10 +261,12 @@ partition_slice reverse_slice(const schema& schema, partition_slice slice);
 
 constexpr auto max_partitions = std::numeric_limits<uint32_t>::max();
+constexpr auto unlimited_tombstones = 0;
 
```
We typically use max() to avoid a special case for 0.

avikivity commented on replica/database.cc:
```diff
@@ -1399,6 +1399,10 @@ bool database::can_apply_per_partition_rate_limit(const schema& s, db::operation
 
+bool database::is_internal_query() const {
+    return classify_request(_dbcfg) != request_class::user;
+}
```
Why make the distinction?

If a system query is paged (as it should be but isn't) it should tolerate empty pages.

If it's unpaged, the query tombstone page limit should be unlimited.

denesb replied:
I wanted to follow the behavior of max result size here: when reached cut the page for paged queries and abort on unpaged ones.

avikivity replied:
Perhaps the internal query client can set a max limit, and avoid this if maze.

denesb replied:
This is much more work, requires accurately tracking down all places initiating internal queries. It is much better to delegate the decision to a single entity (storage proxy), who makes the decision based on the query class. This is the same method we used for max result size.

avikivity commented on idl/read_command.idl.hh:
```diff
@@ -52,6 +52,7 @@ class read_command {
     query::is_first_page is_first_page [[version 2.2]] = query::is_first_page::no;
     std::optional<query::max_result_size> max_result_size [[version 4.3]] = std::nullopt;
     uint32_t row_limit_high_bits [[version 4.3]] = 0;
+    uint64_t tombstone_limit [[version 5.1]] = 0;
```
= max would be more natural.

denesb replied:
ok, will change

avikivity commented on cql3/result_generator.hh:
```diff
@@ -20,6 +20,7 @@ class result_generator {
     lw_shared_ptr<const query::read_command> _command;
     shared_ptr<const selection::selection> _selection;
     cql_stats* _stats;
+    bool _empty;
```
So, our internal client couldn't handle empty pages? Irony

denesb replied:
It can, it currently forwards them to the client, which is actually the intended behavior.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -434,8 +434,14 @@ select_statement::do_execute(query_processor& qp,
+        return p->fetch_page_generator_result(page_size, now, timeout, _stats).then(wrap_result_to_error_message([this, p = p.get()] (result_generator&& generator) -> maybe_result_msg {
+            if (!p->is_exhausted() && generator.empty()) {
+                return std::nullopt; // next page
+            }
```
Shouldn't we return the empty page to the client?

denesb replied:
Turns out this code is not needed indeed. The python client (the one I'm using for tests) handles empty pages well.

nyh replied:
Although the python client can "handle" empty pages, I think if you want to test this code, you want to test individual pages, and see that the first page is empty. Otherwise how does the test test what you want it to test?

denesb replied:
The cql-python tests that I wrote check the query result page-by-page, making sure they have the expected number of empty pages and non-empty-pages.

nyh replied:
Oh, by the way, I forgot, but older clients indeed had bugs with empty pages - and I hve in test/cql-pytest/conftest a fixture "driver_bug_1" to check that the driver is new enough to support empty pages.
In any case, if you want to *check* that the page is empty, I think even older drivers will be fine. You just need to do a query and check that len(results.current_rows) == 0, you don't even need to continue the paging.

denesb replied:
Well, I do want to check that the row I expect is eventually delivered. You can check out my tests in the last commit.

nyh replied:
Good. I just now re-remembered how bad the cassandra-driver support for empty pages is. It was only fixed last year, but there was no new pip release of cassandra-driver in the last year and a half (!), so only scylla-driver also supports empty pages correctly.
That being said, I think it might be possible to write tests that work on the older driver by doing the paging manually and not letting the driver do it, but I think it's not worth the effort - the "driver_bug_1" fixture simply lets you check that the driver is new enough and if it isn't, skip the test.

avikivity commented on mutation_partition.cc:
```diff
@@ -2074,19 +2074,22 @@ void query_result_builder::consume(tombstone t) {
 stop_iteration query_result_builder::consume(range_tombstone_change&& rtc) {
+    _stop = _rb.bump_and_check_tombstone_limit();
     return _stop;
 }
```
Missing: partition_start with tombstone.

We might want to count individual cell tombstones too, but may be too expensive. To we parse cells here to count their size?

avikivity replied:
I guess is_live will be false for rows that are composed of dead cells only.

denesb replied:
Right, I missed the partition tombstone.

Rows which have no content are also considered dead by the compactor, and the will be counted as a tombstone by the `is_live` check in `consume(clustering_row)`.

nyh replied:
> Rows which have no content are also considered dead by the compactor, and the will be counted as a tombstone by the `is_live` check in `consume(clustering_row)`.

I see, makes a lot of sense. If you want to make sure this never gets broken in the future, you can add a test for this specific case (as I suggested above - using UPDATE and DELETE of a single column instead of DELETE of a whole row).

avikivity commented:
Looks good.

nyh commented on query-result-writer.hh:
```diff
@@ -135,6 +138,19 @@ public:
 
+    stop_iteration bump_and_check_tombstone_limit() {
+        ++_tombstones;
+        if (!_tombstone_limit || _tombstones < _tombstone_limit) {
```
Avi is right that if the default tombstone_limit was maxint, this if() would have been simpler.

psarna commented on alternator/executor.cc:
```diff
@@ -1395,7 +1395,8 @@ static lw_shared_ptr<query::read_command> previous_item_read_command(service::st
     auto partition_slice = query::partition_slice(std::move(bounds), {}, std::move(regular_columns), selection->get_query_options());
-    return ::make_lw_shared<query::read_command>(schema->id(), schema->version(), partition_slice, proxy.get_max_result_size(partition_slice));
+    return ::make_lw_shared<query::read_command>(schema->id(), schema->version(), partition_slice, proxy.get_max_result_size(partition_slice),
+            query::tombstone_limit(proxy.get_tombstone_limit()));
```
nitpick: is there any reason why `proxy.get_tombstone_limit()` returns a bare integer instead of this `tombstone_limit` type? It would be more natural to just return the correct type right away

denesb replied:
I guess it made life harder for code using the IDL constructor. Maybe the time came for me to move all code to the new constructor.

denesb commented:
> > Fixes: #7689 Refs: #3672
> 
> Please also check if it fixes or somehow related to the following issues:
> 
>     * [Scan timeout after all data TTL exceeded #3914](https://github.com/scylladb/scylla/issues/3914) Scan timeout after all data TTL exceeded

Should fix this.

> 
>     * [Alternator's Scan should be ok with long strings of tombstones #7933](https://github.com/scylladb/scylla/issues/7933) Alternator's Scan should be ok with long strings of tombstones

Should fix this.

> 
>     * [Long reactor stalls when applying a large number of range tombstones to the cache #10282](https://github.com/scylladb/scylla/issues/10282) Long reactor stalls when applying a large number of range tombstones to the cache

Unrelated. This is related to the cache.

> 
> 
> If any issue is "related" but not fixes, please explain the nature of this relationship and why it doesn't fix - just saying "Refs" doesn't explain enough.

denesb commented on db/config.cc:
```diff
@@ -559,6 +559,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "The maximum number of tombstones a query can scan before warning.")
     , tombstone_failure_threshold(this, "tombstone_failure_threshold", value_status::Unused, 100000,
         "The maximum number of tombstones a query can scan before aborting.")
+    , query_tombstone_limit(this, "query_tombstone_limit", value_status::Used, 10000,
```
ok

denesb commented on mutation_partition.cc:
```diff
@@ -2074,19 +2074,22 @@ void query_result_builder::consume(tombstone t) {
 }
 stop_iteration query_result_builder::consume(static_row&& sr, tombstone t, bool is_live) {
     if (!is_live) {
+        _stop = _rb.bump_and_check_tombstone_limit();
```
I wanted to minimize the diff (and therefore the noise).

denesb commented on mutation_partition.cc:
```diff
@@ -2074,19 +2074,22 @@ void query_result_builder::consume(tombstone t) {
     return _stop;
 }
 stop_iteration query_result_builder::consume(range_tombstone_change&& rtc) {
+    _stop = _rb.bump_and_check_tombstone_limit();
```
`_stop` is always false at the time a `consume()` method is called. As soon as one of the `consume()` methods return false, consumption stops and no more `consume()` methods are called (except a closing `consume_end_of_partition()` and then `consume_end_of_stream()`).

denesb commented on query-result-writer.hh:
```diff
@@ -135,6 +138,19 @@ public:
+        }
+        if (!_slice.options.contains<partition_slice::option::allow_short_read>()) {
+            throw std::runtime_error(fmt::format(
+                    "Tombstones processed by unpaged query exceeds limit of {} (configured via query_tombstone_limit)",
```
Avi suggested query_tombstone_page_limit.

denesb commented on test/boost/cql_query_test.cc:
```diff
@@ -5342,3 +5342,47 @@ SEASTAR_TEST_CASE(test_bind_variable_type_checking) {
+            has_more_pages = ::has_more_pages(result);
+            paging_state = extract_paging_state(result);
+            if (has_more_pages) {
+                assert_that(result).is_rows().with_size(0);
```
Now I'm confused too. I wrote this test before integrating @tgrabiec's empty page support changes, including the loops around paging in `select_statement.cc` which discards empty pages. This test was passing before those changes and kept passing after. I have to look into this.

avikivity replied:
Maybe that work was for internal loops - like when aggregating.

denesb replied:
I will drop this test in favor of one using the external client (python driver). This test will more closely match the end-user experience and hence is more relevant than this test which is using the internal C++ API to fetch pages.

denesb commented:
Force-pushed to save my progress so far. Not all comments are addressed yet, there will be a next version next week, nevertheless, just in case somebody wants to have a look:
New in v2:
* renamed config item query_tombstone_limit -> query_tombstone_page_limit
* use max-int for unlimited case and rename it to max to match partition and row limits
* move all users of read_command IDL constructor to the other, safer constructor
* have `storage_proxy::get_tombstone_limit()` return tagged int
* added handling of mismatched replica continue positions to digest resolver

Not addressed yet:
* count partition tombstones
* empty pages should be forwarded to client
* result_generator::_empty
* more tests

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1482/

avikivity commented:
> * move all users of read_command IDL constructor to the other, safer constructor

You can extract this one too, if it saves you work later.

denesb commented on cql3/result_generator.hh:
```diff
@@ -123,6 +125,10 @@ public:
     }
+
+    bool empty() const {
+        return _empty;
```
Will drop this code in v3.

denesb commented on cql3/statements/select_statement.cc:
```diff
@@ -450,11 +456,16 @@ select_statement::do_execute(query_processor& qp,
 
-    return p->fetch_page_result(page_size, now, timeout).then(wrap_result_to_error_message(
-            [this, p = std::move(p), &options, now](std::unique_ptr<cql3::result_set>&& rs) {
+    return repeat_until_value([this, page_size, now, timeout, &options, p = std::move(p)] () mutable {
```
Will drop in v3.

denesb commented:
New in v3:
* Added comprehensive tests:
    - row tombstone prefix and span
    - range tombstone prefix and span
    - empty rows prefix and span
    - partition tombstones prefix and span
    - static row tombstones prefix and span
* Fixed issues found by tests, including a pre-existing obscure [IDL Frame truncated](https://github.com/scylladb/scylla/issues/9482) that they started to reliably reproduce.
* Implemented stopping on partition tombstones.
* Dropped internal client iterating on empty pages, this is left to the external client (driver), which is the whole point of this exercise: the empty pages act as a sort of heartbeat for a query that would otherwise time out.

TODO:
* multi-node sanity test
* multi-version sanity test

Removing the RFC tag, the branch is production ready pending the sanity tests I mentioned above.

avikivity commented:
The TODO list in the cover letter needs an update.

avikivity commented:
You'll need to rebase past the first commit due to the accident we had with master.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3076,7 +3092,7 @@ class digest_read_resolver : public abstract_read_resolver {
         }
         auto& first = *_digest_results.begin();
-        return std::find_if(_digest_results.begin() + 1, _digest_results.end(), [&first] (query::result_digest digest) { return digest != first; }) == _digest_results.end();
+        return std::find_if(_digest_results.begin() + 1, _digest_results.end(), [&first] (digest_and_last_pos digest) { return digest.digest != first.digest; }) == _digest_results.end();
```
This is introducing an allocation. query::result_digest is just 16 bytes and can be passed in registers, but digest_and_last_pos is a complex type. Better to pass by reference.

denesb replied:
right, will do

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3723,23 +3739,26 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
-                        timeout, digest_algorithm(*_proxy), adjust_rate_limit_for_local_operation(_rate_limit_info));
+                        timeout, digest_algorithm(*_proxy), adjust_rate_limit_for_local_operation(_rate_limit_info)).then([] (rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, full_position>&& r) {
+                return rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, rpc::optional<full_position>>(std::move(r));
+            });
```
This introduces an allocation. Maybe we can change query_result_local_digest() to return the expected type.

denesb replied:
I want to express in the signature of `query_result_local_digest()` that the position is *always* provided by the replica. The `rpc::optional<>` is only added for the RPC layer, to allow for talking to older nodes.

denesb replied:
BTW, why does this introduce an allocation? I use the move constructor of `rpc::tuple`. Is it because of the additional task?

avikivity replied:
Yes, the extra task.

avikivity replied:
> I want to express in the signature of `query_result_local_digest()` that the position is _always_ provided by the replica. The `rpc::optional<>` is only added for the RPC layer, to allow for talking to older nodes.

I agree with that in principle (of course). But this is too much on the hot path and will definitely be measurable (it won't be dominant, but it will be a small regression). I think this case merits breaking the rules. We can run perf_simple_query to verify it.

denesb replied:
Ok, I changed it.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4225,11 +4244,12 @@ result<::shared_ptr<abstract_read_executor>> storage_proxy::get_read_executor(lw
 
-future<rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature>>
+future<rpc::tuple<query::result_digest, api::timestamp_type, cache_temperature, full_position>>
 storage_proxy::query_result_local_digest(schema_ptr s, lw_shared_ptr<query::read_command> cmd, const dht::partition_range& pr, tracing::trace_state_ptr trace_state, storage_proxy::clock_type::time_point timeout, query::digest_algorithm da, db::per_partition_rate_limit::info rate_limit_info) {
```
Seems doable.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5427,9 +5447,25 @@ storage_proxy::handle_read_digest(const rpc::client_info& cinfo, rpc::opt_time_p
+                }
+
+                std::exception_ptr eptr = f.get_exception();
+                if (features().typed_errors_in_read_rpc) {
```
What happened to encode_replica_exception_for_rpc?

denesb replied:
It assumes the appended exception type is the very last in the result tuple, which is only true so long as we cannot change the order of elements in the tuple.

avikivity replied:
@piodul please make a comment

piodul replied:
Maybe we could change encode_replica_exception_for_rpc so that it doesn't append the exception at the end of the tuple, but rather can insert it at a specified position?

denesb replied:
That would solve the problem.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3094,6 +3094,14 @@ class digest_read_resolver : public abstract_read_resolver {
+    rpc::optional<full_position> min_position() {
+        return std::min_element(_digest_results.begin(), _digest_results.end(), [this] (const digest_and_last_pos& a, const digest_and_last_pos& b) {
+            if (!a.last_pos || !b.last_pos) {
+                return (bool(a.last_pos) - bool(b.last_pos)) < 0;
```
gcc 12 will whine, so int(bool(a.last_pos)).

avikivity replied:
also, the subtraction is unneeded, you can compare the bools directly.

What does it mean if one result has a last position and the other doesn't?

denesb replied:
This is only possible in mixed cluster: some replicas don't set last position, while some do. We want to sort the un-engaged position as the smallest one, so it is the position we set on the final result (meaning, there is no replica-provided last-position).

denesb replied:
changed to compare bools

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3995,6 +4003,7 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
                 auto&& [result, digests_match] = res.value();
 
                 if (digests_match) {
+                    result->set_last_position(digest_resolver->min_position());
```
Can we avoid this if we got a full result (common case)?

denesb replied:
Yes, in that case the result can keep the last-position it already has. We still have to do the comparison to check that all replicas set the same last-pos.

denesb replied:
added an if to avoid the copy

avikivity commented on replica/database.cc:
```diff
@@ -1411,6 +1411,10 @@ bool database::can_apply_per_partition_rate_limit(const schema& s, db::operation
+bool database::is_internal_query() const {
+    return classify_request(_dbcfg) != request_class::user;
+}
+
```
Will this be called on the coordinator or replica?

I'm a little worried that if it is called on the replica, different replicas will reach different decisions and we'll get confused.

denesb replied:
Coordinator. This is called by `storage_proxy::get_tombstone_limit()`, which is called by callers of coordinator side `storage_proxy::query()`.

avikivity commented on alternator/auth.cc:
```diff
@@ -133,7 +133,8 @@ future<std::string> get_key_from_roles(service::storage_proxy& proxy, std::strin
-    auto command = ::make_lw_shared<query::read_command>(schema->id(), schema->version(), partition_slice, proxy.get_max_result_size(partition_slice));
+    auto command = ::make_lw_shared<query::read_command>(schema->id(), schema->version(), partition_slice,
+            proxy.get_max_result_size(partition_slice), query::tombstone_limit(proxy.get_tombstone_limit()));
     auto cl = auth::password_authenticator::consistency_for_user(username);
```
Ok, it's called on the coordinator, good. Also saves us from different nodes having different configurations.

It's bad to involve replica::database in a coordinator problem, but we can clean that up later.

denesb replied:
We have the same problem with query max result size. The dilemma is that of where to put the logic of classifying the query, which is needed for both. After some back-and-forth `database` was selected, though it is by no means the perfect choice.

avikivity commented on query-request.hh:
```diff
@@ -318,7 +320,8 @@ public:
                  query::is_first_page is_first_page,
                  std::optional<query::max_result_size> max_result_size,
-                 uint32_t row_limit_high_bits)
+                 uint32_t row_limit_high_bits,
```
(we should have a uint64_t for the row limit, and left the high/low bits for IDL only)

denesb replied:
This is the IDL constructor. Now it is only used by IDL.
The user-facing constructor does have an `uint64_t` (actually a tagged int).

avikivity replied:
Yes

avikivity commented on multishard_mutation_query.cc:
```diff
@@ -841,10 +841,7 @@ class data_query_result_builder {
-            return _res_builder->build(_compaction_state.current_full_position());
-        }
-        return _res_builder->build();
+        return _res_builder->build(_compaction_state.current_full_position());
```
Doesn't it cost us an extra allocation? For the very common case of only having last pages.

denesb replied:
Yes, this will copy a partition and clustering key. I think it is worth it to not have to open the rats-nest of some pages having a last pos and others not.

I will check perf/simple_query to assess the impact.

denesb replied:
> What happens with range tombstones? They can have trailing nulls. Do we encode them correctly?

The type we use for last-pos is `position_in_partition`. It can handle all positions possible with mutation fragments.

> How are pauses after a static row tombstone translated to the client cookie? The client protocol doesn't have the notion of static rows.

Client cookie is just a blob from client's perspective. And since 3131cbea62 we store a `position_in_partition` in it. This reminds me, @nyh can alternator tables have static rows?

avikivity replied:
How does that work? What if a client gets a cookie from a pre-position_in_partition server, and returns it to a post-position_in_partition server?

denesb replied:
Adding full position-in-partition support was implemented by adding a weight and a region field to `paging_state` in addition to the already present clustering-key. These 3 fields is enough to represent a full position-in-partition.
The cookie is therefore forward compatible but not backwards compatible. An old coordinator would misunderstand a new coordinator's cookie, although in a well defined way. But we have a cluster feature barring the usage of position-in-partition values (range tombstones for example) as last-position in paging cookie.

denesb replied:
Although we already have replicas providing their last-position, currently they will only stop at rows, stopping at dead rows/range tombstones is controlled by a cluster feature, so we cannot have non at(clustering key) positions in a mixed cluster.

avikivity replied:
Ok, so no problem then.

avikivity replied:
Indeed I see it in paging_state.idl. Good job to all involved.

nyh replied:
@denesb no, no static rows in Alternator.

avikivity commented on mutation_compactor.hh:
```diff
@@ -420,8 +420,7 @@ public:
         ++_stats.range_tombstones;
-        do_consume(std::move(rtc), consumer, gc_consumer);
-        return stop_iteration::no;
+        return do_consume(std::move(rtc), consumer, gc_consumer);
```
Does this fix a bug?

denesb replied:
I think this can cause a mutation read (read repair) not stop as soon as the memory checker asks it.
I found this because it breaks range-tombstone based limits, but didn't break any existing test.

avikivity replied:
Maybe it wants to be extracted out to be possibly backported. At least move it to a separate patch with an issue, though I don't think it's serious enough for a backport.

denesb replied:
Extracted into https://github.com/scylladb/scylla/pull/11139

avikivity commented on multishard_mutation_query.cc:
```diff
@@ -508,6 +504,23 @@ future<> read_context::save_reader(shard_id shard, const dht::decorated_key& las
+                    }
+                    reader->unpop_mutation_fragment(mutation_fragment_v2(schema, rparts->permit, partition_start(cs.partition_start)));
+                }
+            }
```
Please extract, with testcase.

denesb replied:
I'm not sure I will be able to write a test-case for this. In this series I have a test-case that in theory should reproduce it 100% of the time, yet it reproduces only part of the time. At one point the same test would reproduce it on my laptop, but on my desktop. This suggests it has a timing component though I don't understand why.

avikivity replied:
It's worth to spend some time to understand it, if we have some rare bug lurking somewhere.

denesb replied:
https://github.com/scylladb/scylla/pull/11137
Reproducing the bug doesn't depend on timing after all: in the context of the test I wrote for this PR it depends on the distribution of partitions across vnodes. I used a random key generating method to generate my partition keys, I think it was just a matter of lucky/unlucky seeds.
With a targeted test-case the issue is reproducible 100% of the time.

avikivity replied:
I'm glad I made you chase this even though it probably wasn't fun, I think it's an important bug.

avikivity commented on service/pager/query_pagers.cc:
```diff
@@ -401,7 +401,7 @@ void query_pager::handle_result(
         _exhausted = (row_count < page_size && !results->is_short_read()) || _max == 0;
 
-        if (!_exhausted && row_count > 0) {
+        if (!_exhausted) {
```
What happens in a mixed cluster when talking to old replicas? I guess nothing bad (removed check it redundant), but let's think about it (and document in changelog).

denesb replied:
There are not empty pages here, in an old or mixed cluster. To be honest I don't really know why this check existed here at all. Probably as a sanity check.

avikivity commented on mutation_partition.cc:
```diff
@@ -2074,22 +2074,26 @@ void query_result_builder::consume_new_partition(const dht::decorated_key& dk) {
 }
 stop_iteration query_result_builder::consume(clustering_row&& cr, row_tombstone t,  bool is_live) {
     if (!is_live) {
+        _stop = _rb.bump_and_check_tombstone_limit();
```
Suppose bump_and_check returns no. Won't it override a yes in _stop?

denesb replied:
`_stop` is always `::no` here. After one of the `consume()` methods setting it to `::yes`, the consume loop is immediately broken and `consume_end_of_partition()` is called next.

avikivity replied:
Maybe we shouldn't have a variable _stop then.

avikivity replied:
(e.g. it shoudn't be a member, just a local)

denesb replied:
We need it as a member, because we need to remember whether we stopped do to us (us=`query_result_builder`) requesting the stop, or it was the layer above us. In the former case the read is declared short.

avikivity replied:
I see. Maybe a comment.

denesb replied:
added comment

avikivity commented on query-result-writer.hh:
```diff
@@ -135,6 +138,19 @@ public:
+        }
+        if (!_slice.options.contains<partition_slice::option::allow_short_read>()) {
+            throw std::runtime_error(fmt::format(
+                    "Tombstones processed by unpaged query exceeds limit of {} (configured via query_tombstone_limit)",
```
config item name needs update

denesb replied:
done

denesb commented:
> Does the cookie sent to the client get advanced when we send an empty page? If not, then if the 1865 cache misses, we'll keep looping on the same position.

Yes, the cookie sent to the client contains the last-position as provided by the replica. If the replica doesn't provide it that is a bug that will trigger an assert in the pager.

avikivity commented:
Please run perf_simple_query before/after. I'm curious if the extra allocations I identified have an impact.

denesb commented:
Rebased on fixed master.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1522/

denesb commented:
New in v4:
* Return `rpc::optional<full_position>` from `query_digest_locally()` to avoid an additional task to convert return values on the hot path.
* Only override last-position on query result if: empty-replica-pages feature is set and the min position is smaller than the current one.
* Split out to tangential bugfixes: https://github.com/scylladb/scylladb/pull/11139 and https://github.com/scylladb/scylladb/pull/11137. Both are merged.
* Avoid copy in `digest_read_resolver::digests_match()`.
* Added comment above `query_result_builder::_stop` explaining why we need to remember that we requested to stop.
* Adjusted config item reference in `query_result_builder::bump_and_check_tombstone_limit()`.

Change to query-pagers (removing `_row_count > 0` condition) is safe: there are not empty pages in an old or mixed cluster.

TODO:
* Sanity tests
* Perf tests

denesb commented:
New in v5:
* Fix broken filtering queries.
* Remove incorrect assumption that last-position is always set in new cluster -- it is not set on truly empty pages (when there is no result at all).
* Added test for no-result queries.

TODO:
* More filtering+empty-pages tests.
* Multi-node sanity-test.
* Mixed cluster sanity-test.
* Perf test.

denesb commented:
Upgrade sanity test found that I forgot to update `read_command::tombstone_limit = 0` -> `read_command::tombstone_limit = query::max_tombstones`. This generated a ton of read repairs in a mixed cluster. I guess this is why sanity tests are needed, even if the code was prepared to cope with mixed clusters.

denesb commented:
I fixed the issue above. I'm still not done with my TODO list, I pushed so I can kick off a BYO build. I think for an intrusive change like this a simple CI run is not enough.

denesb commented:
BYO job: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1496/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1670/

denesb commented:
New in v6 (compared to v5):
* rebased
* Fixed wrong default value in read_command IDL
* Fixed segfault in last position propagation
* completed sanity tests

denesb commented:
Upgrade test was conducted as following:
* Created cluster of 3 nodes with RF=3 with master version
* Wrote small dataset of 1000 rows.
* Deleted prefix of 980 rows.
* Started read workload: `scylla-bench -mode=read -workload=uniform -replication-factor=3 -nodes 127.0.0.1,127.0.0.2,127.0.0.3 -clustering-row-count=10000 -duration=10m -rows-per-request=9000 -page-size=100`
* Stopped and upgraded each node one-by-one. New nodes were started by `--query-tombstone-page-limit=10`.
* Confirmed there are no errors or read-repairs.

denesb commented:
Perf test: `build/release/test/perf/perf_simple_query -c1 -m2G --concurrency=100 --duration=5`
Before:
```
138843.29 tps ( 62.1 allocs/op,  12.1 tasks/op,   42359 insns/op,        0 errors)
142232.68 tps ( 62.1 allocs/op,  12.1 tasks/op,   42362 insns/op,        0 errors)
145798.97 tps ( 62.1 allocs/op,  12.1 tasks/op,   42363 insns/op,        0 errors)
141322.78 tps ( 62.1 allocs/op,  12.1 tasks/op,   42350 insns/op,        0 errors)
135312.45 tps ( 62.1 allocs/op,  12.1 tasks/op,   42365 insns/op,        0 errors)
```
After:
```
138843.29 tps ( 62.1 allocs/op,  12.1 tasks/op,   42359 insns/op,        0 errors)
142232.68 tps ( 62.1 allocs/op,  12.1 tasks/op,   42362 insns/op,        0 errors)
145798.97 tps ( 62.1 allocs/op,  12.1 tasks/op,   42363 insns/op,        0 errors)
141322.78 tps ( 62.1 allocs/op,  12.1 tasks/op,   42350 insns/op,        0 errors)
135312.45 tps ( 62.1 allocs/op,  12.1 tasks/op,   42365 insns/op,        0 errors)
```

No material difference.

denesb commented:
Testing the case where one node has a tombstone prefix that the other doesn't revealed the following:
* If the tombstones are not purgeable, the read triggers a read repair, after which both nodes will have the same tombstone prefix.
* If the tombstones are purgeable they won't be counted towards the tombstone limit, so they cause no difference. This can make this series ineffective for a dataset with large amount of purgeable tombstones. But such a data-set can be fixed today already by forcing a compaction.

avikivity commented:
> Perf test: `build/release/test/perf/perf_simple_query -c1 -m2G --concurrency=100 --duration=5` Before:
> 
> ```
> 138843.29 tps ( 62.1 allocs/op,  12.1 tasks/op,   42359 insns/op,        0 errors)
> 142232.68 tps ( 62.1 allocs/op,  12.1 tasks/op,   42362 insns/op,        0 errors)
> 145798.97 tps ( 62.1 allocs/op,  12.1 tasks/op,   42363 insns/op,        0 errors)
> 141322.78 tps ( 62.1 allocs/op,  12.1 tasks/op,   42350 insns/op,        0 errors)
> 135312.45 tps ( 62.1 allocs/op,  12.1 tasks/op,   42365 insns/op,        0 errors)
> ```
> 
> After:
> 
> ```
> 138843.29 tps ( 62.1 allocs/op,  12.1 tasks/op,   42359 insns/op,        0 errors)
> 142232.68 tps ( 62.1 allocs/op,  12.1 tasks/op,   42362 insns/op,        0 errors)
> 145798.97 tps ( 62.1 allocs/op,  12.1 tasks/op,   42363 insns/op,        0 errors)
> 141322.78 tps ( 62.1 allocs/op,  12.1 tasks/op,   42350 insns/op,        0 errors)
> 135312.45 tps ( 62.1 allocs/op,  12.1 tasks/op,   42365 insns/op,        0 errors)
> ```
> 
> No material difference.

Because you copied and pasted before into after...

denesb commented:
> Because you copied and pasted before into after...

Right, fixed (edited).

avikivity commented:
Now it looks like 1000 extra instructions/op. That's 2.4%, not small.

denesb commented:
> Now it looks like 1000 extra instructions/op. That's 2.4%, not small.

Is there something we can do about that? How does one even start looking at where exactly those extra instructions are. I mean the series adds a bunch of code, so it is some of these, but how to find the largest contributor that could maybe attempted to be slimmed down a bit?

avikivity commented:
> > Now it looks like 1000 extra instructions/op. That's 2.4%, not small.
> 
> Is there something we can do about that? How does one even start looking at where exactly those extra instructions are. I mean the series adds a bunch of code, so it is some of these, but how to find the largest contributor that could maybe attempted to be slimmed down a bit?

The first step is to make sure those are real numbers, by running a longer test and increasing task-quota-ms:

```
build/release/test/perf/perf_simple_query -c1 -m2G --concurrency=100 --task-quota-ms 10 --operations-per-shard 20000000
```

Don't run it on a laptop, performance is too unreliable. While clock speed doesn't affect instructions/op directly, it affects batching and that affects instructions/op.

If it reproduces, run it under `perf record` and try `perf diff`.

It's possible that just having more code makes the inliner more shy and we lose its benefit, we may have to just eat the cost.

avikivity commented:
/cc @michoecho for commentary

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1676/

michoecho commented:
> How does one even start looking at where exactly those extra instructions are. I mean the series adds a bunch of code, so it is some of these, but how to find the largest contributor that could maybe attempted to be slimmed down a bit?

@denesb I have learned to do this reliably. This use case is _begging_ for Intel PT, it's a match made in heaven.

You capture before/after PT traces of a query with `perf record`, visualize them (with instruction count as the time axis), put them side by side (scaled to the same unit), and use your eye to find to compare the length of function bars. Then you can drill down on the interesting ones with `perf script` to read (or diff) the exact assembly and source code.

You can find the details at https://github.com/michoecho/perf2perfetto.

This UX leaves a lot to be desired and requires way more manual work and knowledge (familiarity with `perf`'s options) than it could, but I think it's still easily the best tool for the task at the moment.

It could all be automated, including the diffing and drilling down, with a common UI, but unfortunately nobody has done it yet. It's a shame, because this technology has so much potential.

denesb commented:
> @denesb I have learned how to do this reliably. This use case is _begging_ for Intel PT, it's a match made in heaven.

Well, all my CPUs are AMD ones, so I'll try pure perf diff. Alternatively I can use a cloud machine hoping to get an Intel CPU (any tips for provider/instance type?), but that automatically adds hours to the process, as I so rarely have to use them that I always forget even the basics on how to start, connect, deploy, etc.

denesb commented:
```
build/release/test/perf/perf_simple_query -c1 -m2G --concurrency=100 --task-quota-ms 10 --operations-per-shard 20000000
```
If I pass a high value for `--operation-per-shard`, `perf_simple_query` just gets stuck and doesn't print anything for minutes. If I lower the opts count at one point it starts to work, although it runs only for a second or so. The breaking point is between `20000` and `200000`.

denesb commented:
We really need a `docs/dev/benchmarking.md`, as I always forget all the best practices by the time I have to do benchmarks again.
I'll try to put something basic together, then you guys can follow-up with more information. I'll start with collecting the bits and pieces we have scattered in the wiki.

denesb commented:
Rebased on top of master (https://github.com/scylladb/scylladb/pull/11088), perf diff unchanged.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1689/

avikivity commented:
> ```
> build/release/test/perf/perf_simple_query -c1 -m2G --concurrency=100 --task-quota-ms 10 --operations-per-shard 20000000
> ```
> 
> If I pass a high value for `--operation-per-shard`, `perf_simple_query` just gets stuck and doesn't print anything for minutes. If I lower the opts count at one point it starts to work, although it runs only for a second or so. The breaking point is between `20000` and `200000`.

I routinely used 1000000 (for a 5-second run at 200k QPS). I'll try it again and see if we regressed.

avikivity commented:
btw, we should start running these regularly like we do the sstable microbenchmarks (which we ignore since they're unstable).

avikivity commented:
```console
$ build/release/test/perf/perf_simple_query --operations-per-shard 1000000 --smp 1 

...

234783.62 tps ( 62.1 allocs/op,  12.1 tasks/op,   41854 insns/op,        0 errors)

median 234783.62 tps ( 62.1 allocs/op,  12.1 tasks/op,   41854 insns/op,        0 errors)
median absolute deviation: 0.00
maximum: 234783.62
minimum: 234783.62
```

avikivity commented:
With --task-quota-ms I get

```
median 239098.35 tps ( 62.1 allocs/op,  12.1 tasks/op,   41858 insns/op,        0 errors)
```

So it doesn't affect insn/op. I guess not surprising at 239k QPS, where the default concurrency is too low to fill even a 0.5ms task quota.

avikivity commented:
Here is --concurrency 10000:

```
178342.85 tps ( 62.0 allocs/op,  12.0 tasks/op,   42596 insns/op,        0 errors)

median 178342.85 tps ( 62.0 allocs/op,  12.0 tasks/op,   42596 insns/op,        0 errors)
median absolute deviation: 0.00
maximum: 178342.85
minimum: 178342.85
```

Large reduction in QPS. Not too surprising, data cache is probably thrashed.

avikivity commented:
Anyway, I can't reproduce your hang on master. So it's either a regression in your branch, or something more complicated.

avikivity commented:
For the lulz, here's the `perf diff` between --concurrency 100 and 10000. Allocation is faster?! Maybe everything is slower proportionately, but allocation remains the same absolutely (perf diff compares percentages, not absolute time).

```
     7.37%     -2.25%  perf_simple_query_g        [.] seastar::memory::allocate
     2.47%     +1.68%  perf_simple_query_g        [.] replica::table::query
     1.34%     +1.60%  perf_simple_query_g        [.] service::storage_proxy::query_singular
     1.26%     +1.46%  perf_simple_query_g        [.] reader_concurrency_semaphore::maybe_admit_waiters
     0.92%     +0.87%  perf_simple_query_g        [.] dht::tri_compare
     0.75%     +0.78%  perf_simple_query_g        [.] seastar::continuation<seastar::internal::promise_base_with_type<boost::outcome_v2::basic_result<service::storage_proxy_coordinator_query_result, utils::exception_container<exceptions::mutation_write_timeout_exception, exceptions::read_ti>
     2.88%     -0.71%  [vdso]                     [.] 0x00000000000006b0
     0.34%     +0.65%  perf_simple_query_g        [.] schema_registry::get
     1.27%     +0.57%  perf_simple_query_g        [.] compact_radix_tree::tree<cell_and_hash, unsigned int>::node_head::clone<row::row(schema const&, column_kind, row const&)::$_15&>
     0.34%     +0.56%  perf_simple_query_g        [.] seastar::noncopyable_function<seastar::future<void> (reader_permit)>::indirect_vtable_for<replica::database::query(seastar::lw_shared_ptr<schema const>, query::read_command const&, query::result_options, std::vector<nonwrapping_interval<>
     0.33%     +0.51%  perf_simple_query_g        [.] mutation_partition::lower_bound
```

avikivity commented:
Here's perf stat for two runs:

```

 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 100':

          8,777.49 msec task-clock                #    0.778 CPUs utilized          
            12,859      context-switches          #    1.465 K/sec                  
                33      cpu-migrations            #    3.760 /sec                   
             1,786      page-faults               #  203.475 /sec                   
    41,016,866,153      cycles                    #    4.673 GHz                      (72.03%)
       188,538,631      stalled-cycles-frontend   #    0.46% frontend cycles idle     (72.02%)
       434,546,051      stalled-cycles-backend    #    1.06% backend cycles idle      (72.04%)
    86,231,530,444      instructions              #    2.10  insn per cycle         
                                                  #    0.01  stalled cycles per insn  (72.02%)
    15,785,299,318      branches                  #    1.798 G/sec                    (72.05%)
        31,570,935      branch-misses             #    0.20% of all branches          (72.04%)

      11.287694318 seconds time elapsed

       8.646249000 seconds user
       0.074315000 seconds sys
```


vs

```
 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 10000':

         11,966.57 msec task-clock                #    0.814 CPUs utilized          
            12,965      context-switches          #    1.083 K/sec                  
                33      cpu-migrations            #    2.758 /sec                   
             1,814      page-faults               #  151.589 /sec                   
    56,547,221,532      cycles                    #    4.725 GHz                      (71.82%)
       194,658,949      stalled-cycles-frontend   #    0.34% frontend cycles idle     (71.84%)
     3,392,035,056      stalled-cycles-backend    #    6.00% backend cycles idle      (71.87%)
    87,797,756,199      instructions              #    1.55  insn per cycle         
                                                  #    0.04  stalled cycles per insn  (71.85%)
    16,003,941,939      branches                  #    1.337 G/sec                    (71.87%)
        27,333,092      branch-misses             #    0.17% of all branches          (71.87%)

      14.695778581 seconds time elapsed

      11.807550000 seconds user
       0.086160000 seconds sys

```

Large drop in IPC, likely due to cache misses.

avikivity commented:
```
 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 100':

     1,493,638,933      L1-dcache-load-misses                                       

      11.252171363 seconds time elapsed

       8.631313000 seconds user
       0.053685000 seconds sys

```

vs

```
 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 10000':

     1,499,897,860      L1-dcache-load-misses                                       

      14.827214348 seconds time elapsed

      11.994419000 seconds user
       0.055742000 seconds sys

```

Not a huge difference.

avikivity commented:
```
 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 100':

       458,796,268      l2_cache_misses_from_dc_misses                                   

      11.258156829 seconds time elapsed

       8.708784000 seconds user
       0.050776000 seconds sys

```

vs

```
 Performance counter stats for 'build/release/test/perf/perf_simple_query_g --operations-per-shard 2000000 --smp 1 --task-quota-ms 10 --concurrency 10000':

       334,869,898      l2_cache_misses_from_dc_misses                                   

      14.074083677 seconds time elapsed

      11.683813000 seconds user
       0.062803000 seconds sys

```

Fewer? what?

avikivity commented:
```

 Performance counter stats for 'system wide':

        23,319,494      l3_misses                                                   

      11.237051891 seconds time elapsed
```

vs

```
 Performance counter stats for 'system wide':

       423,736,811      l3_misses                                                   

      13.836841661 seconds time elapsed

```

A huge signal. @michoecho (informational)

avikivity commented:
400M misses / 2M ops = 200 misses/op. If a miss is 100ns, then that's 20usec/op. Given an op is about 5sec, most misses were happening in parallel. Magic!

michoecho commented:
@avikivity We already went over this exact sequence of measurements once, about 5-6 months ago. At the time we were wondering why the IPC of perf_simple_query is different than the IPC seen in real (cassandra-stress) tests. And after lots of fooling around with low batching, which wasn't the case, it turned out that perf_simply_query workload simply fits in L3 too neatly. When you increase the perf_simple_query to 10000, you get real-life IPC.
We saw the "faster allocation?!" bit too. I think I even found some believable explanation for it back then, but I have managed to forget it by now, because it wasn't worth remembering.

denesb commented:
```
# Event 'cycles:u'
#
# Baseline  Delta Abs  Shared Object        Symbol                                                                                                                                                                                                           >
# ........  .........  ...................  .................................................................................................................................................................................................................>
#
               +1.00%  perf_simple_query_g  [.] service::abstract_read_executor::execute(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<boost::outcome_v2::basic_result>
               +0.40%  perf_simple_query_g  [.] compound_type<(allow_prefixes)0>::compare
               +0.28%  perf_simple_query_g  [.] full_position::full_position
     0.75%     +0.27%  perf_simple_query_g  [.] compound_type<(allow_prefixes)0>::begin
     0.43%     -0.25%  perf_simple_query_g  [.] service::abstract_read_executor::execute(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<boost::outcome_v2::basic_result>
     5.38%     -0.24%  perf_simple_query_g  [.] seastar::memory::free
     2.42%     -0.23%  perf_simple_query_g  [.] mutation_querier::consume(clustering_row&&, row_tombstone)::$_23::operator()<ser::qr_partition__rows<bytes_ostream> >
               +0.21%  perf_simple_query_g  [.] compound_type<(allow_prefixes)1>::begin
     2.42%     +0.19%  perf_simple_query_g  [.] replica::table::query
     6.72%     -0.18%  perf_simple_query_g  [.] seastar::memory::allocate
```
It looks like there is two things here:
* the copying of `full_position` when taken from the compactor
* comparing the position of all the replicas

denesb commented:
With the following diff:
```diff
07:50 $ git diff ./service
diff --git a/service/storage_proxy.cc b/service/storage_proxy.cc
index e735426afc..c1ce8aff10 100644
--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -3822,12 +3822,12 @@ class digest_read_resolver : public abstract_read_resolver {
         _data_result = foreign_ptr<lw_shared_ptr<query::result>>();
         _digest_results.clear();
     }
-    virtual size_t response_count() const override {
-        return _digest_results.size();
-    }
 public:
     digest_read_resolver(schema_ptr schema, db::consistency_level cl, size_t block_for, size_t target_count_for_cl, storage_proxy::clock_type::time_point timeout) : abstract_read_resolver(std::move(schema), cl, 0, timeout),
                                 _block_for(block_for),  _target_count_for_cl(target_count_for_cl) {}
+    virtual size_t response_count() const override {
+        return _digest_results.size();
+    }
     void add_data(gms::inet_address from, foreign_ptr<lw_shared_ptr<query::result>> result) {
         if (!_request_failed) {
             // if only one target was queried digest_check() will be skipped so we can also skip digest calculation
@@ -3839,7 +3839,7 @@ class digest_read_resolver : public abstract_read_resolver {
             got_response(from);
         }
     }
-    void add_digest(gms::inet_address from, query::result_digest digest, api::timestamp_type last_modified, std::optional<full_position> last_pos) {
+    void add_digest(gms::inet_address from, query::result_digest digest, api::timestamp_type last_modified, std::optional<full_position>&& last_pos) {
         if (!_request_failed) {
             _digest_results.emplace_back(std::move(digest), std::move(last_pos));
             _last_modified = std::max(_last_modified, last_modified);
@@ -4736,7 +4736,7 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
                 auto&& [result, digests_match] = res.value();
 
                 if (digests_match) {
-                    if (exec->_proxy->features().empty_replica_pages) {
+                    if (exec->_proxy->features().empty_replica_pages && digest_resolver->response_count() > 1) {
                         auto& mp = digest_resolver->min_position();
                         auto& lp = result->last_position();
                         if (!mp || bool(lp) < bool(mp) || full_position::cmp(*exec->_schema, *mp, *lp) < 0) {
```
The difference looks like this:
```
median 134007.75 tps ( 62.0 allocs/op,  12.0 tasks/op,   42995 insns/op,        0 errors)
median absolute deviation: 7388.05
maximum: 142120.98
minimum: 109333.43
```
vs.
```
median 128220.69 tps ( 62.0 allocs/op,  12.0 tasks/op,   43157 insns/op,        0 errors)
median absolute deviation: 4857.05
maximum: 137238.61                                   
minimum: 102900.57 
```
diff:
```
# Event 'cycles:u'
#
# Baseline  Delta Abs  Shared Object        Symbol                                                                                                                                                                                                           >
# ........  .........  ...................  .................................................................................................................................................................................................................>
#
     5.64%     -0.89%  perf_simple_query_g  [.] seastar::memory::free
               +0.76%  perf_simple_query_g  [.] service::abstract_read_executor::execute(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<boost::outcome_v2::basic_result>
     0.74%     -0.57%  perf_simple_query_g  [.] service::abstract_read_executor::execute(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<boost::outcome_v2::basic_result>
     3.10%     +0.32%  perf_simple_query_g  [.] service::storage_proxy::query_singular
               +0.28%  perf_simple_query_g  [.] full_position::full_position
     0.97%     +0.28%  perf_simple_query_g  [.] cql3::statements::select_statement::do_execute
     1.07%     +0.28%  perf_simple_query_g  [.] replica::database::query
     1.29%     +0.25%  perf_simple_query_g  [.] reader_concurrency_semaphore::maybe_admit_waiters
     6.10%     -0.25%  perf_simple_query_g  [.] seastar::memory::allocate
     2.59%     +0.24%  perf_simple_query_g  [.] replica::table::query
```
With this the instruction difference is reduced to 200 (from 1000). But this is of course cheating sort-of. The optimization above is disabled by having more than one nodes. It does help CL=ONE queries though, which is a real workload.
The remaining 200 instructions is the copy of `full_position` we make at the end of the page. I'll experiment with dropping the filling of last-position on the last page. Initially I hoped to eliminate the corner case of a page not having a position in a fully upgraded cluster, but I realized I cannot, due to genuinely empty pages (no result at all). So I might as well revert said patch to help small reads.

denesb commented:
Dropping the above mentioned patch brought no change. I think this is as far as we can go. Is this acceptable @avikivity?

avikivity commented:
> @avikivity We already went over this exact sequence of measurements once, about 5-6 months ago. At the time we were wondering why the IPC of perf_simple_query is different than the IPC seen in real (cassandra-stress) tests. And after lots of fooling around with low batching, which wasn't the cause, it turned out that the perf_simply_query workload (edit: data, not text) simply fits in L3 too neatly. When you increase the perf_simple_query concurrency to 10000, you get real-life IPC. We saw the "faster allocation?!" bit too. I think I even found some believable explanation for it back then, but I have managed to forget it by now, because it wasn't worth remembering.

Ah, I remembered that it was text pushing text out. Here it's data pushing text out, not data pushing data?

It's worse on real workloads since --smp 1 gives you the entire L3 (32MiB on my sweet 5950X) for one shard, where in reality it's shared. @bhalevy sharing token_metadata across shards would be good.

avikivity commented:
@denesb I didn't expect this to work with zero increase in instruction count, I just thought 1000 was too much. 200 is acceptable IMO.

avikivity commented:
But if we copy the full_position, how does this not show up in allocation?

avikivity commented:
Why would there be a copy? we reach end-of-stream.

denesb commented:
New in v7:
* Rebased.
* Don't compare positions when there is just one result.
* Drop patch forcing setting a last pos on last page.

```
build/release/test/perf/perf_simple_query_g -c1 -m2G --concurrency=1000 --task-quota-ms 10 --duration=60
```

Before:
```
median 133665.96 tps ( 62.0 allocs/op,  12.0 tasks/op,   43007 insns/op,        0 errors)
median absolute deviation: 973.40
maximum: 135511.63
minimum: 104978.74
```
After:
```
median 129984.90 tps ( 62.0 allocs/op,  12.0 tasks/op,   43181 insns/op,        0 errors)
median absolute deviation: 2979.13
maximum: 134538.13
minimum: 114688.07
```

denesb commented:
> Why would there be a copy? we reach end-of-stream.

For single page requests, there shouldn't be one anymore.

denesb commented:
> But if we copy the full_position, how does this not show up in allocation?

Probably because we have a single page.

avikivity commented:
Ah, I remembered that we use larger keys like cassandra-stress. But I see we use 8-byte keys.

avikivity commented:
Please add the perf report to the cover letter.

avikivity commented:
Also the info about the upgrade test.

denesb commented:
Done.

avikivity commented:
There is a WIP patch there

denesb commented:
> There is a WIP patch there

That is https://github.com/scylladb/scylladb/pull/11249 cherry-picked on top. Will drop it.

denesb commented:
WIP patch dropped.

michoecho commented:
> 400M misses / 2M ops = 200 misses/op. If a miss is 100ns, then that's 20usec/op. Given an op is about 5sec, most misses were happening in parallel. Magic!

@avikivity https://github.com/lemire/testingmlp says that Ryzen 5950X can handle 20-something outstanding cache misses. And a miss is shorter than 100ns (it's 75ns on the tested machine), and an op is also slightly longer than 5µs, so for your measured results you only need ~3 misses in parallel, way below the hardware limits.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1735/

avikivity commented:
Sorry, I forgot - now that 5.1 has been branched, the idl tags should be 5.2.0. It's informational only but better not to be misinformational.

denesb commented:
New in v8:
* Bump version in idl to 5.2

denesb commented:
Speaking of IDL, my `std::optional<full_position` is wrapped by `rpc::optional<>`, ends up being `rpc::optional<std::optional<full_position>>`. This is ugly and redundant but otherwise harmless. Is there any way to instruct IDL to not wrap it? I tried using `rpc::optional<>` directly, but this was wrapped too.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1752/

denesb commented:
New in v9:
* Increase partition-count in tests of partition/static-row tombstone spans. It looks like the current numbers were too low, with certain token configuration the partitions were distributed such that no vnode had a page worth of dead partitions.

avikivity commented:
> Speaking of IDL, my `std::optional<full_position` is wrapped by `rpc::optional<>`, ends up being `rpc::optional<std::optional<full_position>>`. This is ugly and redundant but otherwise harmless. Is there any way to instruct IDL to not wrap it? I tried using `rpc::optional<>` directly, but this was wrapped too.

@gleb-cloudius

gleb-cloudius commented:
> > Speaking of IDL, my `std::optional<full_position` is wrapped by `rpc::optional<>`, ends up being `rpc::optional<std::optional<full_position>>`. This is ugly and redundant but otherwise harmless. Is there any way to instruct IDL to not wrap it? I tried using `rpc::optional<>` directly, but this was wrapped too.
> 
> @gleb-cloudius

Since we are using wonderful 21 century technology (github PRs) I do not get any context here. So what are you trying to do? Any new parameter you add to an existing verb will be wrapped into `rpc::optional `because it is not guarantied it will be present when communicating with older nodes. If this new parameter is in itself optional then `rpc::optional<std::optional<new_param>>` is what you will get. If it is not optional then do not use `std::optional` and you will get `rpc::optional<new_param>`.

denesb commented:
> > > Speaking of IDL, my `std::optional<full_position` is wrapped by `rpc::optional<>`, ends up being `rpc::optional<std::optional<full_position>>`. This is ugly and redundant but otherwise harmless. Is there any way to instruct IDL to not wrap it? I tried using `rpc::optional<>` directly, but this was wrapped too.
> > 
> > 
> > @gleb-cloudius
> 
> Since we are using wonderful 21 century technology (github PRs) I do not get any context here. So what are you trying to do? Any new parameter you add to an existing verb will be wrapped into `rpc::optional `because it is not guarantied it will be present when communicating with older nodes. If this new parameter is in itself optional then `rpc::optional<std::optional<new_param>>` is what you will get. If it is not optional then do not use `std::optional` and you will get `rpc::optional<new_param>`.

Yes, I was adding a new parameter (`std::optional<something>`) to an existing verb and was wondering whether I can avoid it becoming `rpc::optional<std::optional<something>>`. Like I said, this wrapping is not in any way bad, it just looks ugly and if there is an easy way to avoid it I would. If there is not, I won't bother.

gleb-cloudius commented:
> Yes, I was adding a new parameter (`std::optional<something>`) to an existing verb and was wondering whether I can avoid it becoming `rpc::optional<std::optional<something>>`. Like I said, this wrapping is not in any way bad, it just looks ugly and if there is an easy way to avoid it I would. If there is not, I won't bother.

No, you cannot. `rpc::optional<std::optional<something>>` allows you to differentiate between  `something` is optionally was not sent and the case where a sender is old and knows nothing about `something`. You can make `something` to be able to be empty and then you can send it without std::optional<>.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1754/

avikivity commented:
We can periodically create new verbs that don't have all these optional baggage. It's also high time to drop support for old cluster features.

gleb-cloudius commented:
On Wed, Aug 10, 2022 at 03:39:00AM -0700, Avi Kivity wrote:
> We can periodically create new verbs that don't have all these optional baggage. It's also high time to drop support for old cluster features.
> 

To add a new verb you need a new cluster feature :)

--
			Gleb.

avikivity commented:
> On Wed, Aug 10, 2022 at 03:39:00AM -0700, Avi Kivity wrote: We can periodically create new verbs that don't have all these optional baggage. It's also high time to drop support for old cluster features.
> To add a new verb you need a new cluster feature :)


That's fine. But the new verb won't have rpc::optionals and 64-bit ints split into two parts.
